modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientUtils.java

```diff
@@ -495,12 +501,73 @@ void writeObject(BinaryOutputStream out, Object obj) {
 
     /** Read Ignite binary object from input stream. */
     <T> T readObject(BinaryInputStream in, boolean keepBinary) {
-        Object val = marsh.unmarshal(in);
+        if (keepBinary)
+            return (T)marsh.unmarshal(in);
+        else {
+            BinaryReaderHandles hnds = new BinaryReaderHandles();
+
+            return (T)unwrapBinary(marsh.deserialize(in, hnds), hnds);
+        }
+    }
+
+    /**
+     * Unwrap binary object.
+     */
+    private Object unwrapBinary(Object obj, BinaryReaderHandles hnds) {
```

Comment:
This does not look correct to me:

1. Why do we use special unmarshal logic for thin client? It is supposed to be reused from thick, right?
2. We should not need to unwrap things like that. When `keepBinary` is `false`, we should deserialize correctly right away, without intermediate `BinaryObject`s that need to be unwrapped.

Reply:
Well, now we need to decide what should we do with this, considering this is backport from Ignite.

Reply:
I propose to properly fix this in separate patch so it can be donated in Ignite. What do you think?

Reply:
Agree.